images/instrumentation/injector/src/dash0_injector.c

```diff
@@ -82,31 +94,85 @@ char *__getenv(const char *name)
 char val1[1012];
 char val2[1012];
 
-char *getenv(const char *name)
-{
-    char *origValue = __getenv(name);
-    int l = __strlen(name);
-
-    char *nodeOptionsVarName = NODE_OPTIONS_ENV_VAR_NAME;
-    if (__strcmp(name, nodeOptionsVarName) == 0)
-    {
-        if (__strlen(val1) == 0)
-        {
-            // Prepend our --require as the first item to the NODE_OPTIONS string.
-            char *nodeOptionsDash0Require = NODE_OPTIONS_DASH0_REQUIRE;
-            __strcat(val1, nodeOptionsDash0Require);
-
-            if (origValue != NULL)
-            {
-                // If NODE_OPTIONS were present, append the existing NODE_OPTIONS after our --require.
-                __strcat(val1, " ");
-                __strcat(val1, origValue);
-            }
+char *getenv(const char *name) {
+  char *origValue = __getenv(name);
+  int l = __strlen(name);
+
+  char *otelResourceAttributesVarName = OTEL_RESOURCE_ATTRIBUTES_ENV_VAR_NAME;
+  char *nodeOptionsVarName = NODE_OPTIONS_ENV_VAR_NAME;
+  if (__strcmp(name, otelResourceAttributesVarName) == 0) {
+    // If we have not calculated this env var yet
+    if (__strlen(val1) == 0) {
+      char *namespaceName = __getenv(DASH0_NAMESPACE_NAME_ENV_VAR_NAME);
+      char *podUid = __getenv(DASH0_POD_UID_ENV_VAR_NAME);
+      char *podName = __getenv(DASH0_POD_NAME_ENV_VAR_NAME);
+      char *containerName = __getenv(DASH0_POD_CONTAINER_NAME_VAR_NAME);
+
+      int catCount = 0;
+
+      if (namespaceName != NULL && __strlen(namespaceName) > 0) {
+        __strcat(val1, "k8s.namespace.name=");
+        __strcat(val1, namespaceName);
+        catCount += 1;
+      }
+
+      if (podName != NULL && __strlen(podName) > 0) {
+        if (catCount > 0) {
+          __strcat(val1, ",");
         }
 
-        return val1;
+        __strcat(val1, "k8s.pod.name=");
+        __strcat(val1, podName);
+        catCount += 1;
+      }
+
+      if (podUid != NULL && __strlen(podUid) > 0) {
+        if (catCount > 0) {
+          __strcat(val1, ",");
+        }
+
+        __strcat(val1, "k8s.pod.uid=");
+        __strcat(val1, podUid);
+        catCount += 1;
+      }
+
+      if (containerName != NULL && __strlen(containerName) > 0) {
+        if (catCount > 0) {
+          __strcat(val1, ",");
+        }
+
+        __strcat(val1, "k8s.container.name=");
+        __strcat(val1, containerName);
+        catCount += 1;
+      }
+
+      if (origValue != NULL && __strlen(origValue) > 0) {
+        if (catCount > 0) {
+          __strcat(val1, ",");
+        }
+
+        __strcat(val1, origValue);
+      }
     }
 
-    return origValue;
-}
+    return val1;
```

Comment:
_question / double check / rambling:_ According to https://opentelemetry.io/docs/specs/otel/resource/sdk/#specifying-resource-information-via-an-environment-variable all characters outside of "baggage-octet range MUST be percent-encoded."

https://www.w3.org/TR/baggage/#definition:

> baggage-octet          =  %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
>                          ; US-ASCII characters excluding CTLs,
>                          ; whitespace, DQUOTE, comma, semicolon,
>                          ; and backslash

Some SDKs will actually validate the string and crash the process if there are invalid characters:
https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-resources/src/detectors/EnvDetectorSync.ts#L118-L120 ->
https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-resources/src/detectors/EnvDetectorSync.ts#L133-L135

We basically add user provided input here to the env var. We either need to be _very_ sure that none of the sources (pod name, container name, ...) can contain invalid characters or do percent-encoding in the injector.

* namespace names, pod names & container names: I think all of those should fall under the following and thus they should be fine: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names -> lowercase alphanumeric characters, '-' or '.' See also: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L177
* pod uids: obviously those are fine

I'm bringing it up just in case, would you agree with this being safe?

Reply:
This is an excellent point, we should make it safe.

Reply:
After double checking, we concluded that the character set of K8s object names is indeed more restrictive than the baggage-octet, so the current code is safe.